From 1d190149f4126038b3f43081eeb3cf62fc65b7c7 Mon Sep 17 00:00:00 2001
From: OpenEmbedded <oe.patch@oe>
Date: Tue, 25 Apr 2023 14:38:41 +0800
Subject: [PATCH] Fix psu signal and vout issue

Due to mowgli's psu(Artesyn)doesn't support linear-16
conversion-formula, need to modify them by updating
driver patch for fix PSU(Artesyn) read errors.

Signed-off-by: Andy YF Wang <Andy_YF_Wang@wistron.com>
Signed-off-by: LuluTHSu <Lulu_Su@wistron.com>
Signed-off-by: Pat Lin <Pat_Lin@wistron.com>
---
 drivers/hwmon/pmbus/ibm-cffps.c  | 11 ++++++-----
 drivers/hwmon/pmbus/pmbus_core.c |  2 +-
 2 files changed, 7 insertions(+), 6 deletions(-)

diff --git a/drivers/hwmon/pmbus/ibm-cffps.c b/drivers/hwmon/pmbus/ibm-cffps.c
index 9a429d377de1..b20fe1c091a8 100644
--- a/drivers/hwmon/pmbus/ibm-cffps.c
+++ b/drivers/hwmon/pmbus/ibm-cffps.c
@@ -455,8 +455,7 @@ static struct pmbus_driver_info ibm_cffps_info[] = {
 	[cffps1] = {
 		.pages = 1,
 		.func[0] = PMBUS_HAVE_VIN | PMBUS_HAVE_VOUT | PMBUS_HAVE_IOUT |
-			PMBUS_HAVE_PIN | PMBUS_HAVE_FAN12 | PMBUS_HAVE_TEMP |
-			PMBUS_HAVE_TEMP2 | PMBUS_HAVE_TEMP3 |
+			PMBUS_HAVE_PIN | PMBUS_HAVE_FAN12 |
 			PMBUS_HAVE_STATUS_VOUT | PMBUS_HAVE_STATUS_IOUT |
 			PMBUS_HAVE_STATUS_INPUT | PMBUS_HAVE_STATUS_TEMP |
 			PMBUS_HAVE_STATUS_FAN12,
@@ -470,7 +469,7 @@ static struct pmbus_driver_info ibm_cffps_info[] = {
 			PMBUS_HAVE_TEMP2 | PMBUS_HAVE_TEMP3 |
 			PMBUS_HAVE_STATUS_VOUT | PMBUS_HAVE_STATUS_IOUT |
 			PMBUS_HAVE_STATUS_INPUT | PMBUS_HAVE_STATUS_TEMP |
-			PMBUS_HAVE_STATUS_FAN12 | PMBUS_HAVE_VMON,
+			PMBUS_HAVE_STATUS_FAN12,
 		.func[1] = PMBUS_HAVE_VOUT | PMBUS_HAVE_IOUT |
 			PMBUS_HAVE_TEMP | PMBUS_HAVE_TEMP2 | PMBUS_HAVE_TEMP3 |
 			PMBUS_HAVE_STATUS_VOUT | PMBUS_HAVE_STATUS_IOUT,
@@ -519,10 +518,12 @@ static int ibm_cffps_probe(struct i2c_client *client,
 		default:
 		case CFFPS_CCIN_VERSION_1:
 			if ((strncmp(mfg_id, "ACBE", 4) == 0) ||
-				     (strncmp(mfg_id, "ARTE", 4) == 0))
+				     (strncmp(mfg_id, "ARTE", 4) == 0)) {
 				vs = cffps1;
-			else
+			} else {
 				vs = cffps2;
+				ibm_cffps_pdata.flags = PMBUS_SKIP_STATUS_CHECK;
+			}
 			break;
 		case CFFPS_CCIN_VERSION_2:
 			vs = cffps2;
diff --git a/drivers/hwmon/pmbus/pmbus_core.c b/drivers/hwmon/pmbus/pmbus_core.c
index 0f3815583fd9..418e07074fd6 100644
--- a/drivers/hwmon/pmbus/pmbus_core.c
+++ b/drivers/hwmon/pmbus/pmbus_core.c
@@ -635,7 +635,7 @@ static long pmbus_reg2data_linear(struct pmbus_data *data,
 	s32 mantissa;
 	long val;
 
-	if (sensor->class == PSC_VOLTAGE_OUT) {	/* LINEAR16 */
+	if (sensor->class == PSC_VOLTAGE_OUT && !(data->flags & PMBUS_NO_CAPABILITY)) {	/* LINEAR16 */
 		exponent = data->exponent[sensor->page];
 		mantissa = (u16) sensor->data;
 	} else {				/* LINEAR11 */
-- 
2.17.1

